pennylane_honeywell/device.py

```diff
@@ -128,15 +164,184 @@ def set_api_configs(self):
+        """Formats an error message of an HTTP response.
+
+        Args:
+            response: the HTTP response to format
+
+        Returns:
+            str: a formatted version of the response
+        """
+        body = response.json()
+        status = body.get("status_code", "")
+        code = (body.get("code", ""),)
+        detail = (body.get("detail", ""),)
+        meta = (body.get("meta", ""),)
+        return f"{status} ({code}): {detail} ({meta})"
+
+    def _refresh_access_token(self):
+        """Sends a request to refresh the access token using the stored refresh
+        token.
+
+        Returns:
+            str: a valid access token
+
+        Raises:
+            RequestFailedError: if the request failed
+        """
+        # Refresh the access token using the refresh token
+        body = {"refresh-token": self._refresh_token}
+
+        response = requests.post("https://qapi.honeywell.com/v1/login", json=body)
+
+        # Access tokens are also called id-tokens
+
+        if response.status_code == 200:
+            return response.json()["id-token"]
+
+        raise RequestFailedError(
+            f"Failed to get access token: {self._format_error_message(response)}"
+        )
+
+    def get_valid_access_token(self):
+        """Return an access token.
+
+        This method will first try to use any stored tokens (access or refresh
+        token) and otherwise ask for user credentials.
+
+        1. Check the access token:
+            i) if doesn't exist or it expired then check the refresh token
+            ii) otherwise: use it
+
+        2. Check the refresh token:
+            i) if doesn't exist or it expired then ask user for credentials;
+            ii) otherwise: request a new access token using the refresh token
+
+        3. Request a new access token and refresh token using the user
+        credentials
+
+        Returns:
+            str: access token to use for sending requests
+
+        Raises:
+            RequestFailedError: if the request failed
+        """
+        if self._access_token is None or self.token_is_expired(self._access_token):
+
+            if self._refresh_token is None or self.token_is_expired(self._refresh_token):
+                self._access_token, self._refresh_token = self._login()
+                self.save_tokens(self._access_token, refresh_token=self._refresh_token)
+
+            else:
+                # Refresh the access token using the refresh token
+                headers = {"Content-Type": "application/json", "refresh-token": self._refresh_token}
+
+                r = requests.post("https://qapi.honeywell.com/v1/login", json=headers)
+
+                if r.status_code != 200:
+                    raise RequestFailedError(
+                        f"Failed to get access token: {self._format_error_message(r)}"
+                    )
+
+                # Access tokens are also called id-tokens
+                self._access_token = r.json()["id-token"]
```

Comment:
Isn't this basically the same as just calling the `_refresh_access_token` method above? 🤔 
```suggestion
                self._access_token = self._refresh_access_token()
```

Reply:
Oh great catch! Thanks :)